Objects/longobject.c

```diff
@@ -157,170 +214,98 @@ _PyLong_New(Py_ssize_t size)
+        return get_small_int((sdigit)value);
     }
-    result = _PyLong_New(i);
-    if (result != NULL) {
-        Py_SET_SIZE(result, Py_SIZE(src));
-        while (--i >= 0) {
-            result->ob_digit[i] = src->ob_digit[i];
-        }
+
+    if (!fits_in_size_field(value)) {
+        /* FIXME(lpereira): allocate new PyLong with as two or more digits here! See
+         *                  PYLONG_FROM_UINT() macro */
     }
-    return (PyObject *)result;
-}
 
-static PyObject *
-_PyLong_FromMedium(sdigit x)
-{
-    assert(!IS_SMALL_INT(x));
-    assert(is_medium_int(x));
-    /* We could use a freelist here */
     PyLongObject *v = PyObject_Malloc(sizeof(PyLongObject));
     if (v == NULL) {
         PyErr_NoMemory();
         return NULL;
     }
-    Py_ssize_t sign = x < 0 ? -1: 1;
-    digit abs_x = x < 0 ? -x : x;
-    _PyObject_InitVar((PyVarObject*)v, &PyLong_Type, sign);
-    v->ob_digit[0] = abs_x;
-    return (PyObject*)v;
-}
 
-static PyObject *
-_PyLong_FromLarge(stwodigits ival)
-{
-    twodigits abs_ival;
-    int sign;
-    assert(!is_medium_int(ival));
-
-    if (ival < 0) {
-        /* negate: can't write this as abs_ival = -ival since that
-           invokes undefined behaviour when ival is LONG_MIN */
-        abs_ival = 0U-(twodigits)ival;
-        sign = -1;
-    }
-    else {
-        abs_ival = (twodigits)ival;
-        sign = 1;
-    }
-    /* Must be at least two digits */
-    assert(abs_ival >> PyLong_SHIFT != 0);
-    twodigits t = abs_ival >> (PyLong_SHIFT * 2);
-    Py_ssize_t ndigits = 2;
-    while (t) {
-        ++ndigits;
-        t >>= PyLong_SHIFT;
-    }
-    PyLongObject *v = _PyLong_New(ndigits);
-    if (v != NULL) {
-        digit *p = v->ob_digit;
-        Py_SET_SIZE(v, ndigits * sign);
-        t = abs_ival;
-        while (t) {
-            *p++ = Py_SAFE_DOWNCAST(
-                t & PyLong_MASK, twodigits, digit);
-            t >>= PyLong_SHIFT;
-        }
-    }
+    _PyObject_InitVar((PyVarObject *)v, &PyLong_Type, encode_value(value));
     return (PyObject *)v;
 }
 
-/* Create a new int object from a C word-sized int */
-static inline PyObject *
-_PyLong_FromSTwoDigits(stwodigits x)
+PyObject *
+_PyLong_Copy(PyLongObject *src)
 {
-    if (IS_SMALL_INT(x)) {
-        return get_small_int((sdigit)x);
+    PyLongObject *result;
+    Py_ssize_t value;
+
+    assert(src != NULL);
+
+    if (as_inlined_int(src, &value)) {
+        return _PyLong_FromSignedSize(value);
     }
-    assert(x != 0);
-    if (is_medium_int(x)) {
-        return _PyLong_FromMedium((sdigit)x);
+
+    value = Py_ABS(value);
+    result = _PyLong_New(value);
+    if (result == NULL) {
+        PyErr_NoMemory();
+        return NULL;
     }
-    return _PyLong_FromLarge(x);
+
+    /* The size calculation here shouldn't overflow because _PyLong_New() already
+     * checks that value is larger than MAX_LONG_DIGITS. */
+    memcpy(result->ob_digit, src->ob_digit, sizeof(sdigit) * value);
+
+    return result;
```

Comment:
Need `(PyObject *)` cast.

Reply:
A lot of these little things will be ironed out whenever the compiler yells at me.  I still haven't compiled this, so I'm sure I'll hear a lot whenever I'm ready to do that.

Reply:
Well, just understand that I find it easier to review when there aren't little details that could be fixed by trying to compile and fixing what it finds. :-)